network-api/networkapi/buyersguide/tests.py

```diff
@@ -440,18 +460,455 @@ def name_change_changes_slug(self):
+                    '2': 3,
+                    '3': 4,
+                    '4': 5,
+                },
+                'average': 4.0,
+            },
+            'total': 15,
+        }
+        self.assertDictEqual(data, comparable_data)
+
+    def test_get_or_create_votes(self):
+        # Delete potential votes
+        self.product_page.votes.delete()
+        self.product_page.votes = None
+        self.product_page.save()
+        self.assertFalse(hasattr(self.product_page.votes, 'set_votes'))
+
+        votes = self.product_page.get_or_create_votes()
+        self.assertEqual(votes, [0, 0, 0, 0, 0])
+        self.assertTrue(hasattr(self.product_page.votes, 'set_votes'))
+
+
+class WagtailBuyersGuideVoteTest(APITestCase, BuyersGuideTestMixin):
+
+    def test_successful_vote(self):
+        # Reset votes
+        votes = self.product_page.get_or_create_votes()
+        self.product_page.votes.set_votes([0, 0, 0, 0, 0])
+
+        response = self.client.post(self.product_page.url, {
+            'value': 25,
+            'productID': self.product_page.id
+        }, format='json')
+        self.assertEqual(response.status_code, 200)
+
+        self.product_page.refresh_from_db()
+
+        votes = self.product_page.votes.get_votes()
+        self.assertListEqual(votes, [0, 1, 0, 0, 0])
+        self.assertEqual(self.product_page.current_tally, 1)
+
+        response = self.client.post(self.product_page.url, {
+            'value': 100,
+            'productID': self.product_page.id
+        }, format='json')
+        self.assertEqual(response.status_code, 200)
+
+        self.product_page.refresh_from_db()
+
+        votes = self.product_page.votes.get_votes()
+        self.assertListEqual(votes, [0, 1, 0, 0, 1])
+        self.assertEqual(self.product_page.current_tally, 2)
+
+    def test_vote_values_as_strings(self):
+        # Reset votes
+        response = self.client.post(self.product_page.url, {
+            'value': 25,
+            'productID': "string as an id"
+        }, format='json')
+        self.assertEqual(response.status_code, 405)
+
+        response = self.client.post(self.product_page.url, {
+            'value': "twenty five",
+            'productID': self.product_page.id
+        }, format='json')
+        self.assertEqual(response.status_code, 405)
+
+    def test_bad_vote_value(self):
+        # vote = 500
+        response = self.client.post(self.product_page.url, {
+            'value': -1,
+            'productID': self.product_page.id
+        }, format='json')
+        self.assertEqual(response.status_code, 405)
+
+        response = self.client.post(self.product_page.url, {
+            'value': 101,
+            'productID': self.product_page.id
+        }, format='json')
+        self.assertEqual(response.status_code, 405)
+
+    def test_missing_product_vote(self):
+        response = self.client.post(self.product_page.url, {
+            'value': 25,
+            'productID': 9999
```

Comment:
it feels like if we change things so that posting auto-retrieve the product's id instead of pulling it from the JSON, that simplifies both the POST handler, and the test code, making it worth doing now rather than in follow-up.

Reply:
I like it! Will implement now.